modules/storage-api/src/main/java/org/apache/ignite/internal/storage/util/StorageOperation.java

```diff
@@ -25,16 +25,55 @@
  * Storage operations.
  */
 interface StorageOperation {
+    /**
+     * Returns future completion of the operation.
+     */
+    CompletableFuture<Void> operationFuture();
+
+    /**
+     * Return {@code true} if the operation is the final.
+     */
+    boolean isFinalOperation();
+
+    /**
+     * Marks the operation as final.
+     */
+    void markFinalOperation();
+
+    /**
+     * Abstract operation of the storage.
+     */
+    abstract class AbstractStorageOperation implements StorageOperation {
+        private final CompletableFuture<Void> operationFuture = new CompletableFuture<>();
+
+        private volatile boolean finalOperation;
+
+        @Override
+        public CompletableFuture<Void> operationFuture() {
+            return operationFuture;
+        }
+
+        @Override
+        public boolean isFinalOperation() {
+            return finalOperation;
+        }
+
+        @Override
+        public void markFinalOperation() {
+            finalOperation = true;
+        }
+    }
+
     /**
      * Storage creation operation.
      */
-    class CreateStorageOperation implements StorageOperation {
+    class CreateStorageOperation extends AbstractStorageOperation {
```

Comment:
Why are all these classes implemented as internal classes? You can move them to corresponding package.
By the way, why is it located in the "util"?

Reply:
At the moment it is used only in this package, if necessary, we will move it to another package.